### PR TITLE
[ao][sparsity] Composability of fusion and sparsity

### PR DESCRIPTION
**Pull request by HDCharles**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74848
* __->__ #74847

Summary: Similar to the other PRs in this stack, the main problem was
that fusion needed to detect the original module type of parametrized
module when sparse prepare was called before fusion. In addition, there
was a potential issue with fusion before sparse_prepare but after the
sparse_config is created. However, in practice fusion moves the
references to the original modules into the fused module without issue.
Thus the original sparse_config that pointed to the original modules
gets automatically updated. If the fusion method changes this may cause
an issue since no explicit handling or updating of these pointers was
needed.

Test Plan: python test/test_ao_sparsity.py TestComposability

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35240273](https://our.internmc.facebook.com/intern/diff/D35240273)

### CODE REVIEW

File: torch/nn/intrinsic/modules/fused.py [hunk 1]
```diff
@@ -1,6 +1,18 @@
 import torch
 from torch.nn import Conv1d, Conv2d, Conv3d, ReLU, Linear, BatchNorm1d, BatchNorm2d, BatchNorm3d
 
+# from torch.ao.quantization.utils import nonparam_type
+from torch.nn.utils.parametrize import is_parametrized
+def nonparam_type(module):
```
Comment (vkuzo) [inline torch/nn/intrinsic/modules/fused.py:6]: same as previous PR, `type_before_parametrizations`?

Comment (HDCharles) [reply torch/nn/intrinsic/modules/fused.py]: fixed

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240273).

File: torch/nn/intrinsic/modules/fused.py [hunk 2]
```diff
@@ -1,6 +1,18 @@
 import torch
 from torch.nn import Conv1d, Conv2d, Conv3d, ReLU, Linear, BatchNorm1d, BatchNorm2d, BatchNorm3d
 
+# from torch.ao.quantization.utils import type_before_parametrizations
```
Comment (vkuzo) [inline torch/nn/intrinsic/modules/fused.py:4]: remove?

**Review by vkuzo — APPROVED** [review]

File: torch/nn/intrinsic/modules/fused.py (hunk 2 above)
Comment (HDCharles) [reply torch/nn/intrinsic/modules/fused.py:4]: fixed

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240273).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240273).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240273).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240273).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240273).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240273).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240273).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240273).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240273).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35240273).

**Review by jerryzh168 — APPROVED** [review]

**Review by andrewor14 — APPROVED** [review]

**Pull request closed without merging**